refactor(core): unify mesh picking methods

peppsac opened the pull request:
The idea of this PR is to expose a single entry point for querying the view.

This method belongs to the View.

Here are a few example usage:
  * view.pickObjectsAt({ x, y })
  * view.pickObjectsAt({ x, y }, 'wfsBuilding')
  * view.pickObjectsAt({ x, y }, 'wfsBuilding', myLayer)
  * view.pickObjectsAt({ x, y }, 'wfsBuilding', myObject3d)

It supports either layers, layer's id or THREE.Object3D.
By default it'll use THREE.Raycaster, but some layer can provide their own implementation.

Currently the layer with a custom implementation are:
  - layers using TileMesh (globe, planar, panorama)
  - pointclouds

peppsac committed:
refactor(core): unify mesh picking methods

The idea of this PR is to expose a single entry point for querying the view.

This method belongs to the View.

Here are a few example usage:
  * view.pickObjectsAt({ x, y })
  * view.pickObjectsAt({ x, y }, 'wfsBuilding')
  * view.pickObjectsAt({ x, y }, 'wfsBuilding', myLayer)
  * view.pickObjectsAt({ x, y }, 'wfsBuilding', myObject3d)

It supports either layers, layer's id or THREE.Object3D.
By default it'll use THREE.Raycaster, but some layer can provide their own implementation.

Currently the layer with a custom implementation are:
  - layers using TileMesh (globe, planar, panorama)
  - pointclouds

gchoqueux commented on src/Core/Scheduler/Providers/PointCloudProvider.js:
```diff
@@ -130,6 +130,51 @@ function loadPointFile(layer, url) {
     });
 }
 
+function pick(view, mouse, layer) {
```
Good refacto and unification.
However I have an objection on the choice of files for the picking declaration : the providers 

To be more generic, clearer and to reuse,  should be in a picking toolbox.
These tools must have as input the type of object to pick.

Then the layer's picking functionality must be assigned to its construction, but not in the pre-process phase of the provider. Because the provider must take care to supply only what has in title and pre-process works only for this supply.

it's same case in PR #546 
what is your motivation to put this in the providers?

I also ask the opinion to  @mbredif ?

peppsac replied:
> To be more generic, clearer and to reuse, should be in a picking toolbox.
> [...]
>  what is your motivation to put this in the providers?

Yes; we already had this discussion earlier, so I won't copy-paste all my arguments from there.

I can move `pickObjectsAt` to a `Picking.js` file if it you prefer.

But I still believe that `preprocessing` of geometry Providers is a good place to extend layer's object so I'll leave the `layer.pickObjectsAt` there.

peppsac replied:
Done.

gchoqueux commented on src/Core/Scheduler/Providers/TileProvider.js:
```diff
@@ -9,6 +9,31 @@ import TileGeometry from '../../TileGeometry';
+import RendererConstant from '../../../Renderer/RendererConstant';
+import Picking from '../../Picking';
+
+function changeRenderState(tileLayer) {
```
All references to the layer should disappear from `tileProvider`. Because there isn't reason for a tile provider to know the layers. For example, in the future, remove any notion of layer in `TileProvider.prototype.executeCommand`. You have to be able to use the tile provider without a layer concept. 

`preprocessDataLayer` also has no reason to be here and even less if that add functions that aren't `tile building`

So don't add new reverence to the layer

autra replied:
> Because there isn't reason for a tile provider to know the layers

Every single provider of itowns absolutely *does* know about layers. Providers are here to interface layers with 2 things: 
- a protocol by:
    - initializing the layer with some default values for this protocol and / or metadata from the network, following the protocol specifications (getCapabilities, tileset.json...)
    - abstract the fetching of data from servers implementing this protocol
- The scheduler mechanism, because we're fetching datas from the internet. And it uses the layer because it provides some configurations that `executeCommand` needs to generate new nodes (opacity, colors, overrideMaterials etc...)

And in turns, the way we create node for a given protocol have an influence on how we should pick it on the scene. So it makes a lot of sense that this logic should be at least partly in providers.

> preprocessDataLayer also has no reason to be here

For the aforementioned reason, it has all the reason to be in a provider

Not sure I understand you well. Could you describe what alternative solution you would choose? Where would you move all that code then? It might make things clearer.

peppsac replied:
Moved more code so now TileProvider isn't modified.

Is this what you had in mind @gchoqueux ?

gchoqueux replied:
The best place for picking is the construction of the geometric layer using the three.js raycaster caster picking. And then overload it in more specific cases like the tile layer. In this way, there is no need to add the picking in each provider of geometry.
@autra  It's clearer?

autra replied:
Somehow... 

> And then overload it in more specific cases like the tile layer

Where would you overload it then? I found that providers was a good place for that (because they constructs the objects, so they know how they should be picked). That being said I'm fine with the current version.

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,98 @@
+}
+
+export default {
+    pickTilesAt: (_view, mouse, layer) => {
```
The picking function shouldn't have as parameter the layer but the objects to pick.
Because this function must be usable without the notion of layer

gchoqueux commented on src/Core/View.js:
```diff
@@ -458,4 +458,106 @@ View.prototype.execFrameRequesters = function execFrameRequesters(when, dt, upda
+ * view.pickObjectsAt({ x, y }, 'wfsBuilding')
+ * view.pickObjectsAt({ x, y }, 'wfsBuilding', myLayer)
+ */
+View.prototype.pickObjectsAt = function pickObjectsAt(mouse, ...where) {
```
why `pickObjectsAt` isn't found in picking.js

peppsac replied:
Because Picking.js is the implementation of the various picking methods (I'm also moving `raycastObjectsAt` in Picking.js)

`View.prototype.pickObjectsAt` is the boilerplate needed to use the pickings methods on a view.

peppsac replied:
See previous answer.

It's boilerplate code to ease picking usage from the View.

gchoqueux replied:
ok

gchoqueux commented on src/Core/Scheduler/Providers/PointCloudProvider.js:
```diff
@@ -160,6 +161,7 @@ export default {
         layer.update = PointCloudProcessing.update;
         layer.postUpdate = PointCloudProcessing.postUpdate;
 
+        layer.pickObjectsAt = (view, mouse) => Picking.pickPointsAt(view, mouse, layer);
```
`pickObjectsAt` must be moved in geometry layer

peppsac replied:
(small aside: please avoid this wording. "You must do this" is kinda rude)

There's no perfect place for this kind of code actually. PR #546 aimed to fix this but it couldn't be merged yet...

Here you're nitpicking on one line and your argument "Provider is to provide things" can be used to not move the picking code in `Layer.js` : "Layer should only contain generic things but Picking.pickPointsAt is a custom pick code".

So I suggest we move and work on improving #546 (or submit another PR that would solve this problem).

gchoqueux replied:
It can have several providers return the same type of object, THREE.Object3D for example, it makes more sense to move it in the construction of the geometric layer than to put picking code in  each provider that returns an object3D

peppsac replied:
> put  picking code in each provider that returns an object3D

- I'm doing this solely for the PointCloudProvider
- this line will go away as soon as #546 is merged
- I disagree that putting this pointcloud specific line is better in GeometryLayer constructor.

Will you block this PR for this?

gchoqueux replied:
Are you ok with this [comment](https://github.com/iTowns/itowns/pull/635#discussion_r166295422)?

Otherwise, I agree, it makes more sense, first of all, to work and merge #546

peppsac replied:
No.
I'd prefer to first merge this one, and then merge the more complicated one.

peppsac replied:
Reminder: this PR doesn't remove nor add feature. It simply groups the picking-from-window-coordinate in the same module. I don't even see why we're arguing so much about it :(

gchoqueux replied:
add a temporary comment : to indicate that it'll not stay in the provider

peppsac commented:
BRanch updated.

gchoqueux commented:
There is also `FeatureUtils` there is also foo as a method of picking

peppsac commented:
> There is also FeatureUtils there is also foo as a method of picking

`FeatureUtils.filterFeaturesUnderCoordinate` is picking from a `Coordinates` object while this PR is focusing on objects selection from mouse coordinates.

Later, it could be interesting to extend the picking API to any kind of coordinates (pixels  or geographic) though.

gchoqueux commented:
Without picking feature, so idea to expose a single entry point for querying isn't completed?

peppsac commented:
> Without picking feature, so idea to expose a single entry point for querying isn't completed?

You can pick features at given window coordinates: `view.pickObjectsAt(mouse, wfsLayer)` will return all features under your mouse (or you can use Picking.pickObjectsAt(mouse, camera, someObject3d)` if you prefer, but for wfs the first form is easier since wfs objects are attached to TileMesh).

gchoqueux commented:
and for 2D vector features?

peppsac commented:
> and for 2D vector features?

This function can pick anything that is rendered as a Mesh. Does that answer your question?

gchoqueux commented:
yes, the PR title and commit should be changed to : `unify picking mesh methods`

peppsac commented:
Added a small change : `Picking` module also returns the `layer.id` of the object (if it exists).

peppsac commented:
So can I rebase this PR (and change the title) or are there any more comments?

peppsac commented:
Rebased. And added a small fix: make sure we only display objects from the layer when doing the picking (see https://github.com/iTowns/itowns/pull/635/files#diff-11ecfe7e737c10e75920ac54de872437R8)

peppsac commented:
Rebased since #637 was merged.

peppsac commented:
@gchoqueux can you update your review of this PR please? (or dismiss it if you don't have time to update your review)

gchoqueux commented:
there's [comment](https://github.com/iTowns/itowns/pull/635#discussion_r166301777) without reply 
and I add this previous [comment](https://github.com/iTowns/itowns/pull/637#issuecomment-364384095)

peppsac commented:
> there's comment without reply

I've answered to all comments in this thread 6 days ago so...
If I missed something please be more specific.

>  and I add this previous comment

which I also answered.

gchoqueux commented:
> I disagree that putting this pointcloud specific line is better in GeometryLayer constructor.

Why?

peppsac commented:
> Why

Already answered as well in the comment  [above](https://github.com/iTowns/itowns/pull/635#discussion_r165957424): `Layer.js` should only contains generic code that applies to all layers, and `layer.pickObjectsAt = (view, mouse) => Picking.pickPointsAt(....)` is pointcloud generic.
And a more important argument: IMHO this dicussion is ±pointless because PR #546 should reorganize the code properly.
And see also the issue #182 that details the need for refactoring regarding Providers.

So unless you feel this is really important and you intend to block this PR because of this, I consider this a non issue.

gchoqueux commented on examples/3dtiles.html:
```diff
@@ -152,15 +152,12 @@
-                raycaster.setFromCamera(mouse, globe.camera.camera3D);
-                // calculate objects intersecting the picking ray
-                var intersects = raycaster.intersectObjects($3dTilesLayerRequestVolume.object3d.children, true);
+                const intersects = view.pickObjectsAt(
```
i'll be better than that goes on like this:
```js
layer.on('mouseMove', (event) => { 
    const intersects = event.intersects;
}
```

peppsac replied:
Feel free to implement `.on()` for layers if you think it's useful, but that's not the scope of this PR.

gchoqueux commented:
With approach `viewer.on`

* remove methods, `eventToViewCoords,  eventToNormalizedCoords, viewToNormalizedCoords and normalizedToViewCoords` in itowns's core and all users cases. Because their responsibilities are unloaded to the `viewerEvent`
* the same event `viewerEvent` is created and is dispatched to all listeners (which avoids redundant calculations)
* simplifies the code

```js
function picking(event) {
        var intersects = view.pickObjectsAt(view.eventToViewCoords(event), layer);
}
window.addEventListener('mousemove', picking, false);
```
become
```js
 function picking(event) {
        var intersects = event.intersects;
 }
 view.on('mousemove', picking, layer);
```
-------------------------------

**An example of partial implementation**
```js
const viewerListeners = {};
const raycaster = new Raycaster();
function viewerEvent(type, event, viewer) {
    const _event = {
        type,
        viewCoords: _eventCoords.set(event.offsetX, event.offsetY),
        event,
    };
    let normalizedCoords;
    let intersects;
    Object.defineProperty(
        _event,
        'normalizedCoords',
        {
            get: () => {
                if (!normalizedCoords) {
                    _eventCoords.x = 2 * (_event.viewCoords.x / viewer.camera.width) - 1;
                    _eventCoords.y = -2 * (_event.viewCoords.y / viewer.camera.height) + 1;
                    normalizedCoords = _eventCoords;
                }
                return normalizedCoords;
            },
        });
    Object.defineProperty(
    _event,
    'intersects',
        {
            get: () => {
                if (!intersects) {
                    if (_event.any instanceof GeometryLayer) {
                        raycaster.setFromCamera(_event.normalizedCoords, viewer.camera.camera3D);
                        intersects = raycaster.intersectObjects(_event.any.object3d.children, true);
                    } else {
                        intersects = [];
                    }
                }
                return intersects;
            },
        });

    return _event;
}

View.prototype.on = function _on(type, listener, any) {
    if (!viewerListeners[type]) {
        const domElement = this.mainLoop.gfxEngine.renderer.domElement;
        viewerListeners[type] = (event) => {
            this.dispatchEvent(viewerEvent(type, event, this));
        };
        if (type == 'click') {
            domElement.addEventListener('mousedown', viewerListeners[type]);
        } else if (type == 'mousemove') {
            domElement.addEventListener('mousemove', viewerListeners[type]);
        }
    }
    this.addEventListener(type, (event) => {
        event.any = any || { object3d: this.scene };
        listener(event);
    });
};
```

peppsac commented:
Your proposed implementation won't work properly for pointclouds (raycasting in millions of points doesn't work well).
You want to replace the line:
```js
   intersects = raycaster.intersectObjects(_event.any.object3d.children, true);
```
With:
```js
   intersects  = _event.any.pickOjectsAt(viewer, viewer.eventToViewCoords(_event));
```
To get something that works with all types of GeometryLayers (and of course `pickOjectsAt` is exactly the function added by this PR).

And again, sorry but your `view.on(...)` proposal isn't relevant for this PR (already said so [earlier](https://github.com/iTowns/itowns/pull/635#discussion_r167595186)).

gchoqueux commented:
The `view.on (...)` approach doesn't prevent appropriate processing for each type of `GeometryLayer`
Just define it in the `intersects` getter property

autra commented:
Another PR for `view.on` please. We need to measure all the impacts with a working implementation (I'm not against it in itself btw). It's largely orthogonal with this changeset.

peppsac commented:
@gchoqueux you missed the most important part of my comment : "And again, sorry but your view.on(...) proposal isn't relevant for this PR (already said so earlier).".
Maybe the solution here is not the best but it's an improvment over the current code.

gchoqueux commented:
I didn't miss it. I will study this PR, so that it can be easily improved towards the proposal `view.on()`.

autra commented:
@gchoqueux so you're blocking a PR for a proposal that is:
- unrelated to it
- won't even touch the same line numbers

why?

peppsac commented:
PR rebased and commits squashed.
@gchoqueux there's nothing here that will prevent implementing the `.on(...)` mechanism.

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,159 @@
+
+    restore.forEach(r => r());
+
+    var depthRGBA = new THREE.Vector4().fromArray(buffer).divideScalar(255.0);
```
declare `depthRGBA` outside the function

peppsac replied:
Done.

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,159 @@
+
+    return () => {
+        view.scene = oldScene;
+        oldParent.add(object);
```
In cloud globe example, `oldParent` could be null on first picking

peppsac replied:
This shouldn't be possible? Can you be more specific?

gchoqueux replied:
in pointcloud_globe.html example, in the first double click, `oldParent` is null

gchoqueux commented on src/Core/View.js:
```diff
@@ -509,4 +510,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+        }
+    }
+
+    for (let i = 0; i < source.length; i++) {
```
could be replace by `for (const s of source)`?

gchoqueux commented on src/Core/View.js:
```diff
@@ -509,4 +510,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+        this.getLayers(l => l.type == 'geometry') :
+        [...where];
+
+    for (let i = 0; i < source.length; i++) {
```
you convert all the layer'id string to layer?

peppsac replied:
Yes.

gchoqueux commented on src/Core/View.js:
```diff
@@ -509,4 +510,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+                });
+
+                const obj = [];
+                // raycast using parent layer object3d
```
by example, if `source[i]`is point cloud layer the `raycast` method isn't appropriate

peppsac replied:
That's why we don't do raycast. We use instead the specialized `pickObjectsAt` method (see 10 lines above).

gchoqueux replied:
ok

gchoqueux commented on src/Core/View.js:
```diff
@@ -509,4 +510,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+
+                const obj = [];
+                // raycast using parent layer object3d
+                Picking.pickObjectsAt(
```
~~we could add layers filter parameter in the parent's method `pickObjectsAt`~~

peppsac replied:
Sorry I don't understand your comment.

gchoqueux replied:
forget the previous comment

Why geometry layer doesn't have default `pickObjectsAt` method?

peppsac replied:
All GeometryLayer have a default picking method (see https://github.com/iTowns/itowns/pull/635/files#diff-61e012c3e00bdc510be33d9eac684b2aR80)

But the method in View support picking for all objects (for instance raw THREE.Object3D) so we need to use `Picking.pickObjectsAt` as a fallback.

gchoqueux replied:
> for instance raw THREE.Object3D 

Should be explain in the title commit because it doesn't unify the picking position/coordinates and features (in raster)

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,159 @@
+    pickObjectsAt(mouse, camera, object, target = []) {
+        // raycaster use NDC coordinate
+        const ndc = {
+            x: 2 * (mouse.x / camera.width) - 1,
```
`View.prototype.viewToNormalizedCoords `could be a static method and used here (and avoid redundant code)

peppsac replied:
Instead I'll change the signature, to match the layer pickObjectsAt function:
```js
pickObjectsAt(view, mouse, object, target = []) { ... }
```

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,159 @@
+    return Math.round(unpack);
+}
+
+function findLayerIdInParent(obj) {
```
Could you used `Object3D.traverseAncestors` ?

peppsac replied:
No, because `traverseAncestors` can't be stopped as soon as you find what you were looking for.

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,159 @@
+const raycaster = new THREE.Raycaster();
+
+export default {
+    pickTilesAt: (_view, mouse, layer) => {
```
it should also move `getPickingPositionFromDepth` in picking.js

peppsac replied:
Well no.
For now these functions are only valid for tile based layer - so they'll stay in their respective View implementation until we provide a way to query this for ~all geometry layers.

gchoqueux replied:
`pickTilesAt` is only valid for tile based layer?

peppsac replied:
yep

gchoqueux replied:
your proposed picking allows you to get position for which type of geometry layers?

gchoqueux commented on src/Core/Picking.js:
```diff
@@ -0,0 +1,159 @@
+
+        // render 1 pixel
+        // TODO: support more than 1 pixel selection
+        const buffer = view.mainLoop.gfxEngine.renderViewTobuffer(
```
may be change function parameters to `renderViewTobuffer(scene, camera, x, y, width, height)` to clarify the scene switch in the viewer (and avoid `fakeScene`)

peppsac replied:
Let's discuss the change to this function in the dedicated PR: https://github.com/iTowns/itowns/pull/672

In the meantime this PR can use the fakeScene trick because it's harmless.

peppsac replied:
I simplified this code and removed the need for the `fakeScene`

gchoqueux commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -363,8 +365,8 @@ GlobeView.prototype.removeLayer = function removeImageryLayer(layerId) {
+    const picked = this.wgs84TileLayer.pickObjectsAt(this, mouse);
+    const selectedId = picked.length ? picked[0].object.id : undefined;
 
     for (const n of this.wgs84TileLayer.level0Nodes) {
```
it is necessary to browse the tree so we have the node in `picked [0] .object`?

peppsac replied:
I'm not sure to understand what you mean.
But we have to browse the tree to select/deselect tiles based on the picking result.

gchoqueux replied:
in other PR: it will later be necessary to keep the selection to avoid traverse the tree to update the nodes

gchoqueux commented on src/Core/TileMesh.js:
```diff
@@ -81,9 +86,30 @@ TileMesh.prototype.changeState = function changeState(state) {
+    }
+
+    const oldState = this._state;
+    this.traverse(n => applyChangeState(n, state));
```
i don't know whether to check if `n` is an instance of `TileMesh`?

peppsac replied:
It's not needed since `applyChangeState` checks for `changeState` existence.

gchoqueux replied:
yes

gchoqueux commented on src/Core/View.js:
```diff
@@ -512,4 +513,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+        this.getLayers(l => l.type == 'geometry') :
+        [...where];
+
+    for (let i = 0; i < sources.length; i++) {
```
can be a single loop is sufficient?

peppsac replied:
It could be done in a single loop but I think it helpes readability to separate the 2 loops.

gchoqueux replied:
With a comment will be more readable
and could you replace by `for (const source of sources)`

peppsac commented:
I think I've answered/fixed to all your comments. Did I miss something?

gchoqueux commented on src/Core/View.js:
```diff
@@ -512,4 +513,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+                    }
+                });
+
+                const obj = [];
```
use of a target isn't necessary

```js
const obj = Picking.pickObjectsAt(
                    this,
                    mouse,
                    parentLayer.object3d);
```

peppsac replied:
Fixed both comments.

peppsac replied:
And also adapted pickObjectsAt to accept either mouse coords or a (Mouse|Touch)Event

gchoqueux commented on src/Core/View.js:
```diff
@@ -512,4 +513,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
 };
 
+/**
+ * Return objects from some layers/objects3d under the mouse in this view.
```
returns only object3D?

peppsac replied:
See the @return documentation for the returned values.

peppsac commented:
I'll change the commit title when squashing the commits.

gchoqueux commented on src/Core/View.js:
```diff
@@ -512,4 +513,82 @@ View.prototype.normalizedToViewCoords = function normalizedToViewCoords(ndcCoord
+            } else {
+                //   - it hasn't: this layer is attached to another one
+                let parentLayer;
+                this.getLayers((l, p) => {
```
if a geometric layer is attached to another geometric layer.
We could affect it the reference of `parentLayer.object3d` in the `GeometryLayer.prototype.attach` function
`layer.object3d = this.object3d`

Who would avoid the party with `it hasn't: this layer is attached to another one`

WDYT?

peppsac replied:
In other places of itowns, layer.object3d is the object used to add layer's mesh to the scene.

So I don't think it make sense for now to store parent layer's object3D here.

gchoqueux commented:
OK rebase, squash and re word commit then we merge

peppsac commented:
Done

gchoqueux reviewed (approved):
thanks
